Be strict about endpoint creation with undeclared numeric ports

sttts opened the pull request:
The Mesos cloud provider depends on the declaration of ports in the pod
spec in order to create port forwardings from the container IP to the host. In the case of the Mesos endpoint controller these host ports are then used as the endpoints for Kubernetes services.

The old endpoint controller code only checks that named port in service specs are actually existing in the selected pods. If not, an endpoint is not created for a given pod. 

In contrast endpoints for numeric ports are blindly created, independently whether it is declared in the pod spec or not.

This PR deprecates undeclared port by showing a "DEPRECATION NOTICE" in the endpoint controller logs.

Fixes mesosphere/kubernetes-mesos#365

sttts committed:
Announce webserver port in NodePort e2e test

In k8s's default networking model container ports in pod specs are not necessary
in order to get an endpoint for a given service via the endpoint controller. In
other setups like Mesos this information is essential to allow the creation of
necessary port forwardings.

Fixes mesosphere/kubernetes-mesos#365: "Services should be able to create a functioning NodePort service",

sttts committed:
Be strict about endpoint creation with undeclared numberic ports

Cloud providers (like e.g. Mesos) depend on the declaration of ports in the pod
spec in order to create proper port forwarding for endpoints. The old code only
checked that for named port, but allowed undeclared numeric ports.

sttts commented:
This one only includes the changes to the test. Might be less controversial.

sttts committed:
Set container port protocol in endpoints_controller_test because it's checked now

The protocol was unset before this patch. But because numeric targetPorts of
services were blindly accepted for endpoints, this did not matter.

sttts committed:
Don't find undeclared ports in endpoints_controller_test

FindPort should ignore ports which are undeclared in the pods.

sttts committed:
Fix endpoint ports in endpoints_controller_test to be those declared in the test pods

The service port 8088 was not declared, but 8081 was. So this patch changes the
former to the later.

sttts commented:
/cc @jdef

jdef commented:
I like the strictness of this approach. Not sure why this wasn't being validated before. Possibly an upstream use case that I'm not aware of: I don't know the value of a service targetPort referencing a numerical container port that's not actually defined in the pod. some special, magical default behavior that's defined somewhere else?

/cc @thockin

thockin commented:
This is and has always been allowed because, like docker's default mode, declaring ports is advisory not compulsory.  If we change this now, it will surely break things.  I'm disinclined to make this change just before a milestone - is it sufficient to just document this requirement on Mesos?

sttts commented:
Of course this can be documented. But I have the strong feeling that being non-strict is not a good idea.
Some reasons:

- pods become less portable between network implementations, maybe not only Mesos will require the container port information one day, also thinking about firewalls, static system models, network analysis
- relying on undeclared container properties leads to "magic" behavior behind the scenes without additional value
- strict ports prevent errors because in a non-strict k8s selecting pods for a service without a listening process on the tragetPort will lead to connection errors
- it is easy to provide container ports as a user
- it is hard to repair once 1.0 is out

It's pretty much like allowing untyped terms in a typed system: don't do it if you don't have very very good reasons.

thockin commented:
I am generally a fan of struct, but the risk of destabilizing
lord-knows-what is VERY real.  I'd rather do something like announce that
undeclared ports are deprecated, log them (but don't get too spammy) and
give people (us) time to move off them.

On Mon, Jun 29, 2015 at 11:35 PM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> Of course this can be documented. But I have the strong feeling that being
> non-strict is not a good idea.
> Some reasons:
>
>    - pods become less portable between network implementations, maybe not
>    only Mesos will require the container port information one day, also
>    thinking about firewalls, static system models, network analysis
>    - relying on undeclared container properties leads to "magic" behavior
>    behind the scenes without additional value
>    - strict ports prevent errors because in a non-strict k8s selecting
>    pods for a service without a listening process on the tragetPort will lead
>    to connection errors
>    - it is easy to provide container ports as a user
>    - it is hard to repair once 1.0 is out
>
> It's pretty much like allowing untyped terms in a typed system: don't do
> it if you don't have very very good reasons.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/10395#issuecomment-117016448>
> .
>

sttts commented:
sure, I of course understand that risk so near to a release. I like the idea to warn about deprecation. Let me find a good way to output those warnings, probably in the endpoint controller, without repeating the warning hundreds of times.

sttts committed:
Deprecate undeclared numeric ports when creating an endpoint

When the endpoints of a service change and a pod with undeclared numeric ports
is found among those selected by the service, a deprecation warning is printed
in the endpoint controller. The endpoints for those pods are still created for
now.

sttts commented:
The upper commit spams as little as possible without actually remembering any state like pod UIDs in the endpoint controller or comparing new pods with the existing endpoints. Now you get one deprecation message *per service port* and *per endpoint* set change.

sttts committed:
Document deprecation of endpoints without explicitly exposed ports

sttts commented:
Rebased to latest master

thockin commented on docs/api.md:
```diff
@@ -41,6 +41,7 @@ Changes to services are the most significant difference between v1beta3 and v1.
 * The `service.spec.portalIP` property is renamed to `service.spec.clusterIP`.
 * The `service.spec.createExternalLoadBalancer` property is removed. Specify `service.spec.type: "LoadBalancer"` to create an external load balancer instead.
 * The `service.spec.publicIPs` property is deprecated and now called `service.spec.deprecatedPublicIPs`. This property will be removed entirely when v1beta3 is removed. The vast majority of users of this field were using it to expose services on ports on the node. Those users should specify `service.spec.type: "NodePort"` instead. Read [External Services](services.md#external-services) for more i ...
+* Endpoints for pods which do not explicitly expose numeric ports are deprecated. In a future version the endpoint controller will create endpoints for those pods anymore, in analogy to the case of named ports which are not defined for a pod, but referenced from a service.
```
s/will/will not/

sttts replied:
Fixed.

thockin commented:
I can live with this.

@bgrant0607 for API consideration

This is non-critical and can wait until after 1.0 I think.

thockin commented:
LGTM, hold until after 1.0

sttts committed:
Fix documentation typo

sttts commented:
@thockin how should we go forward with this? To rebase I would like to add the deprecation notice to the next api version. But I guess this is pretty far away.

jdef commented:
needs rebase

jdef commented on pkg/service/endpoints_controller.go:
```diff
@@ -304,7 +305,20 @@ func (e *EndpointController) syncService(key string) {
+
 			portNum, err := findPort(pod, servicePort)
+
+			// handle DEPRECATED undeclared numeric ports gracefully for now.
```
does our mesos endpoint controller do this already (log the warning?) or fail? I can't find a warning in that code so I'm assuming that it fails.

sttts replied:
I think it just does not find the referenced port and no endpoint is created.

bgrant0607 commented on pkg/service/endpoints_controller.go:
```diff
@@ -352,6 +366,15 @@ func (e *EndpointController) syncService(key string) {
+	// of the selected pods of the service
+	if len(podsByUndeclaredNumericPorts) > 0 {
+		for port := range podsByUndeclaredNumericPorts {
+			glog.Warningf("DEPRECATION NOTICE: the service %s/%s points to pods which do not explicitly declare the numeric port %v", service.Namespace, service.Name, port)
```
This warning is not useful. Users won't see it.

bgrant0607 commented on docs/api.md:
```diff
@@ -41,6 +41,7 @@ Changes to services are the most significant difference between v1beta3 and v1.
 * The `service.spec.portalIP` property is renamed to `service.spec.clusterIP`.
 * The `service.spec.createExternalLoadBalancer` property is removed. Specify `service.spec.type: "LoadBalancer"` to create an external load balancer instead.
 * The `service.spec.publicIPs` property is deprecated and now called `service.spec.deprecatedPublicIPs`. This property will be removed entirely when v1beta3 is removed. The vast majority of users of this field were using it to expose services on ports on the node. Those users should specify `service.spec.type: "NodePort"` instead. Read [External Services](services.md#external-services) for more i ...
+* Endpoints for pods which do not explicitly expose numeric ports are deprecated. In a future version the endpoint controller will not create endpoints for those pods anymore, in analogy to the case of named ports which are not defined for a pod, but referenced from a service.
```
Even if we decide to allow this change, it's not a change in the v1 API. It's just a non-backward-compatible change at this point.

sttts replied:
This paragraph was phrased before v1 was final I think. Of course, it has to go somewhere else, v1.1 or whatever will be the next step.

bgrant0607 commented on docs/services.md:
```diff
@@ -76,6 +76,8 @@ deploying and evolving your `Services`.  For example, you can change the port
 number that pods expose in the next version of your backend software, without
 breaking clients.
 
+Also note that endpoints for pods without explicitly exposed numeric ports are deprecated. Hence, in order to address pods via a service the container ports – even the numeric ones – must be explicitly declared in the specification of the pod. Future versions of Kubernetes will not create endpoints otherwise and the pods will not be made available through services.
```
We've talked about adding port ranges to services in the future. Would that also be a problem for mesos?

sttts replied:
As long as they are explicit and we can do the right thing in the executor to launch pods with those ranges, it should be fine.

thockin replied:
We considered making the default for services be identity-mapped - meaning 1-65535 -> 1-65535.  I imagine this would be a problem...

bgrant0607 commented:
This is a non-backward-compatible behavioral change, and one that would make the system harder to use.

Today, users don't need to explicitly declare ports on their pods if they don't reference the ports by name from other parts of the API. Docker users find that to be convenient. If we were to change that, we'd probably need something to extract exposed ports from images automatically.

mikedanese commented:
IIUC, this is a only problem if port-mapping has to be configured at container creation time. Perhaps we should focus on removing that constraint rather than adding constraints to the networking model?

bgrant0607 commented:
I don't understand how this improves the situation for Kubernetes-Mesos at all. Either way, if a user didn't declare ports on their pods, the service will not work.

thockin commented:
I really don't think this is a great idea, but worth putting out there...

Add `pod.status.containers[*].ports`.  Make kubelet populate that with info from the container image (EXPOSE).  Make loops like this merge the declared list and the detected list.

sttts commented:
... and it assumes that ports are actually exposed by container images. I have seen so many images which don't do that.

thockin-cc commented:
If the container does not EXPOSE the port and the pod does not declare it,
you guys are up a creek.  Requiring pods to declare ports has downsides - I
am trying to find a solution with less downside.
On Aug 18, 2015 1:27 AM, "Dr. Stefan Schimanski" <notifications@github.com>
wrote:

> ... and it assumes that ports are actually exposed by container images. I
> have seen so many images which don't do that.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/10395#issuecomment-132118961>
> .
>

bgrant0607 commented:
I don't know what problem this is trying to solve.

thockin commented:
Moving towards declared ports has the upside that we can block undeclared
ports (something our security team would like).  It has the downside of
adding friction to users.  I was brainstorming ideas on how to do this
without friction.  I do not really want to add a cluster-global flag (like
docker's --icc flag), but that could be a last resort.  We have a very old
bug open to read EXPOSEd ports from docker images, also.

On Tue, Aug 18, 2015 at 2:06 PM, Brian Grant <notifications@github.com>
wrote:

> I don't know what problem this is trying to solve.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/10395#issuecomment-132350911>
> .
>

thockin commented:
Hey all,

This PR seems to have stalled.  What should we do with it?  A hangout to discuss?  A f2f?

bgrant0607 commented:
I don't think we can restrict to declared ports without adding support for port ranges, up to and including all ports.

jdef commented:
@sttts is on vacation through the end of the week, returning this coming
Monday. I'd like to suggest a hangout once he's back.

On Tue, Aug 25, 2015 at 2:05 PM, Brian Grant <notifications@github.com>
wrote:

> I don't think we can restrict to declared ports without adding support for
> port ranges, up to and including all ports.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/10395#issuecomment-134688152>
> .
>

thockin commented:
@jdef can you arrange a hangout meeting?

On Tue, Aug 25, 2015 at 5:13 PM, James DeFelice <notifications@github.com>
wrote:

> @sttts is on vacation through the end of the week, returning this coming
> Monday. I'd like to suggest a hangout once he's back.
>
> On Tue, Aug 25, 2015 at 2:05 PM, Brian Grant <notifications@github.com>
> wrote:
>
> > I don't think we can restrict to declared ports without adding support
> for
> > port ranges, up to and including all ports.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
> https://github.com/kubernetes/kubernetes/pull/10395#issuecomment-134688152
> >
>
> > .
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/10395#issuecomment-134776248>
> .
>

bgrant0607 commented:
This is a behavioural change that would break backward-compatibility of the v1 API. This change cannot be made without allowing existing clients/applications to work as they do now.
